paddle/operators/shrink_state_op.cc

```diff
@@ -0,0 +1,154 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+#include "paddle/framework/lod_rank_table.h"
+#include "paddle/operators/array_operator.h"
+#include "paddle/operators/math/math_function.h"
+
+namespace paddle {
+namespace operators {
+
+class ShrinkStateOp : public ArrayOp {
```

Comment:
I don't think `ShrinkStateOp` is a good name. We don't really 'shrink' the memory block, we only do a slice on the original one. And `state` is also confusing. Maybe we can call it `RearrangeRnnMemoryOp`

Reply:
We should indicate `RNN` in the name since the Op will only be used in RNN. We could even put `DyRNN` in the name.

Reply:
Done.